File: WPF Databinding in Visual Studio 2010 RTM/[C#]-WPF Databinding in Visual Studio 2010 RTM/C#/WpfEfSampleApp/MyBaseCollection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections.ObjectModel;
using System.Data;
using WpfEfDAL;

namespace WpfEfSampleApp
{
    public abstract class MyBaseCollection<T> : ObservableCollection<T>
    {
        private Boolean _isLoading = false;
        public Boolean IsLoading
        {
            get {
                return _isLoading;
            }
        }

        public Boolean HasChanges
        {
            get {
                IEnumerable<System.Data.Objects.ObjectStateEntry> changed =
                        _context.ObjectStateManager.GetObjectStateEntries(  EntityState.Added |
                                                                            EntityState.Deleted |
                                                                            EntityState.Modified);

                IEnumerable<System.Data.Objects.ObjectStateEntry> entityChanged = from o in changed
                                                                     where (o.Entity is T)
                                                                     select o;

                
                return ((entityChanged != null) && (entityChanged.Count() > 0));
            }
        }

        private OMSEntities _context;
        public OMSEntities Context
        {
            get {
                return _context;
            }
        }

        public MyBaseCollection(IEnumerable<T> query, OMSEntities context)
        {
            _isLoading = true;
            _context = context;
            foreach (T o in query)
            {
                this.Add(o);
            }
    
            _isLoading = false;
        }
    }
}

File: WPF Databinding in Visual Studio 2010 RTM/[VB]-WPF Databinding in Visual Studio 2010 RTM/VB/WpfEfSampleApp/MyBaseCollection.vb
Imports System.Collections.ObjectModel
Imports System.Data
Imports WpfEfDAL

Public MustInherit Class MyBaseCollection(Of T)
    Inherits ObservableCollection(Of T)

    Private _isLoading As Boolean = False
    Public ReadOnly Property IsLoading() As Boolean
        Get
            Return _isLoading
        End Get
    End Property

    Public ReadOnly Property HasChanges() As Boolean
        Get
            Dim changed = _context.ObjectStateManager.GetObjectStateEntries( _
                                                      EntityState.Added Or _
                                                      EntityState.Deleted Or _
                                                      EntityState.Modified)

            Dim entityChanged = From o In changed Where TypeOf o.Entity Is T
            Return (entityChanged IsNot Nothing AndAlso entityChanged.Count > 0)
        End Get
    End Property

    Private _context As OMSEntities
    Public ReadOnly Property Context() As OMSEntities
        Get
            Return _context
        End Get
    End Property

    Protected Sub New(ByVal query As IEnumerable(Of T), ByVal context As OMSEntities)
        _isLoading = True
        _context = context

        If query IsNot Nothing Then
            For Each o In query
                Me.Add(o)
            Next
        End If
        _isLoading = False
    End Sub

End Class
